File: r/theoryforge/R/examples.R
#' Access to the example theories shipped inside the package.
#'
#' Thin wrappers over \code{system.file()}, so that the two twins name the
#' examples the same way: the Python package exposes \code{example_path()} and
#' \code{example_names()} over the same files. The copies under
#' \code{inst/fixtures/} are written by \code{scripts/gen_golden.py} and gated in
#' CI, so they cannot drift from the originals at the repository root.
#' @name examples
#' @keywords internal
NULL

#' Names of the packaged example files
#'
#' The bundled set is four example theories and one literature corpus,
#' \code{panic-corpus.yaml}, which [tf_read_corpus()] reads.
#'
#' @return A sorted character vector of the \code{.yaml} file names. The Python
#'   twin's \code{example_names()} applies the same extension filter, so the two
#'   list the same set.
#' @examples
#' tf_example_names()
#' @export
tf_example_names <- function() {
  dir <- system.file("fixtures", package = "theoryforge")
  if (!nzchar(dir)) {
    stop("could not locate the packaged example theories", call. = FALSE)
  }
  sort(list.files(dir, pattern = "\\.yaml$"))
}

#' Path to a packaged example file
#'
#' @param name File name, e.g. \code{"panic-network.theory.yaml"}. See
#'   [tf_example_names()].
#' @return The path to the installed file.
#' @examples
#' tf_read(tf_example_path("panic-network.theory.yaml"))$id
#' @export
tf_example_path <- function(name) {
  path <- system.file("fixtures", name, package = "theoryforge")
  if (!nzchar(path)) {
    stop("no packaged example named '", name, "'; available: ",
         paste(tf_example_names(), collapse = ", "), call. = FALSE)
  }
  path
}

File: python/src/theoryforge/examples.py
"""Access to the example theories shipped inside the package.

The R twin reaches the same files through ``system.file("fixtures", ...)``, so a
reader who installed only one half still has something to run. The files are
copies of ``fixtures/*.yaml`` at the repository root, written by
``scripts/gen_golden.py`` and gated in CI, so they cannot drift.
"""
from __future__ import annotations

from importlib.resources import as_file, files
from pathlib import Path


def example_names() -> list[str]:
    """The names of the example theory and corpus files, sorted."""
    return sorted(p.name for p in (files("theoryforge") / "fixtures").iterdir())


def example_path(name: str) -> Path:
    """Filesystem path of a packaged example, e.g. ``panic-network.theory.yaml``.

    Raises ``FileNotFoundError`` naming the available files when ``name`` is not
    one of them.
    """
    resource = files("theoryforge") / "fixtures" / name
    if not resource.is_file():
        raise FileNotFoundError(f"no packaged example named {name!r}; available: {example_names()}")
    # A wheel is a directory on disk once installed, so this resolves to a real
    # path; as_file() keeps the call correct for a zipped install too.
    with as_file(resource) as path:
        return Path(path)
